data_iterator.go

```diff
@@ -107,81 +127,133 @@ func (d *DataIterator) Run(tables []*TableSchema) {
 					}
 
 				}
-
-				logger.Debug("table iteration completed")
-
-				// Right now the BatchWriter.WriteRowBatch happens synchronously in
-				// this method. If it ever becomes async, this MarkTableAsCompleted
-				// call MUST be done in WriteRowBatch somehow.
-				d.StateTracker.MarkTableAsCompleted(table.String())
 			}
 		}()
 	}
 
-	i := 0
-	loggingIncrement := len(tablesWithData) / 50
-	if loggingIncrement == 0 {
-		loggingIncrement = 1
-	}
+	for _, key := range d.paginationKeys {
+		tableName := key.table.String()
+		stateBatches, loadBatchesFromState := d.StateTracker.batchProgress[tableName]
+
+		if loadBatchesFromState {
+			for batchIndex, batch := range stateBatches {
+				if batch.Completed {
+					continue
+				}
+
+				batchQueue <- &PaginationKey{
+					table:              key.table,
+					startPaginationKey: batch.LatestPaginationKey,
+					endPaginationKey:   batch.EndPaginationKey,
+					batchIndex:         batchIndex,
+				}
+			}
+		} else {
+			// Set start to minus one since cursor is searching for greater values
+			tableStartPaginationKey := key.startPaginationKey - 1
+			tableEndPaginationKey := key.endPaginationKey
+
+			// Number of batches are set to number of processes, unless each batch becomes smaller than the cursor size
+			keyInterval := tableEndPaginationKey - tableStartPaginationKey
+			concurrencyBatchSize := math.Ceil(float64(keyInterval) / float64(d.Concurrency))
+			batchSize := uint64(math.Max(concurrencyBatchSize, float64(d.CursorConfig.BatchSize)))
+
+			d.logger.WithFields(logrus.Fields{
+				"table":              tableName,
+				"batchSize":          batchSize,
+				"endPaginationKey":   tableEndPaginationKey,
+				"startPaginationKey": tableStartPaginationKey,
+			}).Debugf("queueing %d batches", (keyInterval/batchSize)+1)
+
+			for batchStartPaginationKey := tableStartPaginationKey; batchStartPaginationKey < tableEndPaginationKey; batchStartPaginationKey += batchSize {
+				batchEndPaginationKey := batchStartPaginationKey + batchSize
+				batchIndex := (batchStartPaginationKey - tableStartPaginationKey) / batchSize
+
+				// Set batchEndPaginationKey to endPaginationKey if out of bounds
+				if batchEndPaginationKey > tableEndPaginationKey {
+					batchEndPaginationKey = tableEndPaginationKey
+				}
+
+				paginationKey := &PaginationKey{
+					table:              key.table,
+					startPaginationKey: batchStartPaginationKey,
+					endPaginationKey:   batchEndPaginationKey,
+					batchIndex:         batchIndex,
+				}
 
-	for table, _ := range tablesWithData {
-		tablesQueue <- table
-		i++
-		if i%loggingIncrement == 0 {
-			d.logger.WithField("table", table.String()).Infof("queued table for processing (%d/%d)", i, len(tablesWithData))
+				d.StateTracker.RegisterBatch(tableName, paginationKey.batchIndex, paginationKey.startPaginationKey, paginationKey.endPaginationKey)
+				batchQueue <- paginationKey
+
+				// Protect against uint64 overflow, this might happen if the table is full
```

Comment:
Not sure if I quite follow this. I think I can figure it out with some effort, but maybe we can just illustrate this via an example in the comment as opposed to making the reader figure it out?